snap/gadget.go

```diff
@@ -234,6 +256,231 @@ func ReadGadgetInfo(info *Info, classic bool) (*GadgetInfo, error) {
+	}
+
+	if s == "" {
+		return errors.New(`type is not specified`)
+	}
+
+	if s == "bare" {
+		// unknonwn blob
+		return nil
+	}
+
+	if s == "mbr" {
+		// backward compatibility for type: mbr
+		return nil
+	}
+
+	var isGPT, isMBR bool
+
+	idx := strings.IndexRune(s, ',')
+	if idx == -1 {
+		// just ID
+		switch {
+		case validTypeID.MatchString(s):
+			isMBR = true
+		case validGUUID.MatchString(s):
+			isGPT = true
+		default:
+			return fmt.Errorf("invalid format")
+		}
+	} else {
+		// hybrid ID
+		code := s[:idx]
+		guid := s[idx+1:]
+		if len(code) != 2 || len(guid) != 36 || !validTypeID.MatchString(code) || !validGUUID.MatchString(guid) {
+			return fmt.Errorf("invalid format of hybrid type")
+		}
+	}
+
+	if schema != "gpt" && isGPT {
+		// type: <uuid> is only valid for GPT volumes
+		return fmt.Errorf("GUID structure type with non-GPT schema %q", vol.Schema)
+	}
+	if schema != "mbr" && isMBR {
+		return fmt.Errorf("MBR structure type with non-MBR schema %q", vol.Schema)
+	}
+
+	return nil
+}
+
+func validateRole(vs *VolumeStructure, vol *GadgetVolume) error {
+	if vs.Type == "bare" {
+		if vs.Role != "" && vs.Role != "mbr" {
+			return fmt.Errorf("conflicting type: %q", vs.Type)
+		}
+	}
+	vsRole := vs.Role
+	if vs.Type == "mbr" {
+		// backward compatibility
+		vsRole = "mbr"
+	}
+
+	switch vsRole {
+	case "system-data":
+		if vs.Label != "" && vs.Label != "writable" {
+			return fmt.Errorf(`role of this kind must have an implicit label or "writable", not %q`, vs.Label)
+		}
+	case "mbr":
+		if vs.Size > 446 {
+			return errors.New("mbr structures cannot be larger than 446 bytes")
+		}
+		if vs.Offset != 0 {
+			return errors.New("mbr structure must start at offset 0")
+		}
+		if vs.ID != "" {
+			return errors.New("mbr structure must not specify partition ID")
+		}
+		if vs.Filesystem != "" && vs.Filesystem != "none" {
+			return errors.New("mbr structures must not specify a file system")
+		}
+	case "system-boot", "":
+		// noop
+	default:
+		return fmt.Errorf("unsupported role")
+	}
+	return nil
+}
+
+func validateBareContent(vc *VolumeContent) error {
+	if vc.Source != "" || vc.Target != "" {
+		return fmt.Errorf("cannot use non-image content for bare file system")
+	}
+	if vc.Image == "" {
+		return fmt.Errorf("missing image file name")
+	}
+	return nil
+}
+
+func validateFilesystemContent(vc *VolumeContent) error {
+	if vc.Image != "" || vc.Offset != 0 || vc.OffsetWrite != (GadgetRelativeOffset{}) || vc.Size != 0 {
+		return fmt.Errorf("cannot use image content for non-bare file system")
+	}
+	if vc.Source == "" || vc.Target == "" {
+		return fmt.Errorf("missing source or target")
+	}
+	return nil
+}
+
+func validateStructureUpdate(up *VolumeUpdate, vs *VolumeStructure) error {
+	if vs.IsBare() && len(vs.Update.Preserve) > 0 {
+		return errors.New("preserving files during update is not supported for non-filesystem structures")
+	}
+
+	names := make(map[string]bool, len(vs.Update.Preserve))
+	for _, n := range vs.Update.Preserve {
```

Comment:
This pattern repeats, perhaps factor out a helper?

```
validateUniqueStrings(items []string) error  {
    seen. := make(map[string]bool, len(items))
    for _, item := rangeg items {
      if seen(item) {
          return fmt.Errorf("duplicate element %q",  item)
      }
      seen[item] = true
    }
    return nil
}
```

Reply:
I'm not sure the other cases involve a []string directly, so it would get a bit awkward, also we would need to send in fragments to get a good error message